README
NOTE: use the latest Github version of mojolicious, as code is changing quickly!


MojoX Scaffold Summary
---------------------------

In many cases, you have to create the same template and route structures and controller methods again and again.

For example, in order to manage a list of users, you need routes, controller methods and templates to
    - to list all users,
    - to display information on a single user and
    - to create, update and delete users.

MojoX Scaffold helps you to set up a controller, template files and routes with a few commands.


MojoX Scaffold in action
---------------------------

1.) Install MojoX Scaffold, using cpanminus, this is just

    cpanm https://github.com/forwardever/mojox-scaffold

2.) Create a Mojolicious app:

    mojolicious generate app Test

3.) Switch to the newly created Test app:

    cd Test

4.) Set the correct namespace

Information on routes can now be found in the Test.pm file in the /lib folder.
Now set the desired namespace for the controller files. As we want all of our
controllers to be located under

    /lib/Test/Controllers (namespace Test::Controllers)

we add

    $self->routes->namespace("Test::Controllers");

to the beginning of the startup method in the routes file.

Also remove the existing sample routes!

5.) Create a resource named "users"

Make sure that you are still in the Test folder (NOT the lib folder).

    perl script/test generate scaffold

Enter the required information step by step.

At the end, some manual corrections are required.

Add to your startup method (in case you used dbi):

    ### Plugins
    $self->plugin('resourceful_routes');
    $self->plugin('model_instance', namespace => 'Test::DBI', method => 'dbh');

    ### Resources
    $self->resources('users');

Make sure that you put the correct config data in the database BASE file:
    lib/Test/DBI.pm

and also add a table to your database (there will be an sql command generated
for this which works fine on mysql, but might have to be modified on other db systems)


$self->resources('users') automatically generates the following routes
on each start of your app:

        # GET /users/new - form to create a user
        $r->route('/users/new')->via('get')
          ->to(controller => 'users', action => 'new_form')
          ->name('users_new_form');

        # GET /users/123 - show user with id 123
        $r->route('/users/:id')->via('get')
          ->to(controller => 'users', action => 'show')->name('users_show');

        # GET /users/123/edit - form to update a user
        $r->route('/users/:id/edit')->via('get')
          ->to(controller => 'users', action => 'edit_form')
          ->name('users_edit_form');

        # GET /users - list of all users
        $r->route('/users')->via('get')
          ->to(controller => 'users', action => 'index')
          ->name('users_index');

        # POST /users - create new user
        $r->route('/users')->via('post')
          ->to(controller => 'users', action => 'create')
          ->name('users_create');

        # PUT /users/123 - update an existing user
        $r->route('/users/:id')->via('put')
          ->to(controller => 'users', action => 'update')
          ->name('users_update');

        # DELETE /users/123 - delete an existing user
        $r->route('/users/:id')->via('delete')
          ->to(controller => 'users', action => 'delete')
          ->name('users_delete');


7.) Start your app:

    perl script/test daemon --reload

    Also start your database (e.g. mysql).

8.) Enter the following URL to see the result:

    http://localhost:3000/users

9.) Note: as PUT and DELETE methods are not supported by old HTML standard, you
have to add a hidden parameter in your update and delete forms:
<input type="hidden" name="_method" value="put" />
<input type="hidden" name="_method" value="delete" />


Changes
0.04    2010-11-03 00:00:00
        - Added step-by-step config
        - Added dbi support
        - Added object-db support
        - Removed pure perl support

0.03    2010-10-26 00:00:00
        - Improved template scaffolding (tag helpers)
        - Improved controller scaffolding (working model using a perl class array)
        - Added test scaffolding
        - prepared option to allow scaffolding for various ORMs 


0.02    2010-10-19 00:00:00
        - Added scaffolding (mojolicious generate resource appname resourcename)
        - Removed namespace option (now: admin-users generates
          $r->route("/admin/users") ...
        - Added docs to README


0.01    2010-10-18 00:00:00
        - First release


lib/Mojolicious/Plugin/ResourcefulRoutes.pm
package Mojolicious::Plugin::ResourcefulRoutes;

use strict;
use warnings;

use base 'Mojolicious::Plugin';

use Mojo::ByteStream;

our $VERSION = '0.04';

sub register {
    my ($self, $app) = @_;

    $app->plugins->add_hook(
        before_dispatch => sub {
            my ($self, $c) = @_;

            if ($c->req->method eq 'POST') {
                if (lc($c->req->param('_method')) eq 'delete') {
                    $c->req->method('DELETE');
                }
                elsif (lc($c->req->param('_method')) eq 'put') {
                    $c->req->method('PUT');
                }
            }
        }
    );

    $app->helper(
        resource => sub {
            my ($c, $resource, $options) = @_;
            die
              'ResourcefulRoutes: options have to be passed as hash ref, use'
              . ' "resources" (plural) to define multiple resources at once:'
              . ' resource name: '
              . $resource
              . ', passed options: '
              . $options
              unless ref $options eq 'HASH' || !$options;
            $self->generate_routes($c, $resource, $options);
        }
    );

    $app->helper(
        resources => sub {
            my $c      = shift;
            my @params = @_;

            while (@params) {
                my $resource = shift(@params);
                my $options  = shift(@params)
                  if @params && ref $params[0] eq 'HASH';
                $options ||= {};
                $self->generate_routes($c, $resource, $options);
            }
        }
    );
}

sub generate_routes {
    my $self = shift;
    my ($c, $resource, $options) = @_;

    my $singular   = $options->{singular};


    # Create path for routes
    my $path = join('/', split(/-/, $resource) );


    # Resource name is part of the route name
    my $name = $resource;


    # Create controller path
    my $ctrl = $resource;


    # Get routes object
    my $r = $c->app->routes;


    # Singular resource, i.e. app knows id value (e.g. from login)
    if ($singular) {
        
        my $nr = $r->route("$path");

        # GET /article/new - form for create an article
        $nr->route('/new')->via('get')
          ->to(controller => $ctrl, action => 'new_form')
          ->name($name . '_new_form');

        # GET /article - show article
        $nr->route('/')->via('get')
          ->to(controller => $ctrl, action => 'show')->name($name . '_show');

        # GET /article/edit - form for update an article
        $nr->route('/edit')->via('get')
          ->to(controller => $ctrl, action => 'edit_form')
          ->name($name . '_edit_form');

        # GET /article/delete - form to confirm delete
        $nr->route('/delete')->via('get')
          ->to(controller => $ctrl, action => 'delete_form')
          ->name($name . '_delete_form');

        # POST /article - create article
        $nr->route('/')->via('post')
          ->to(controller => $ctrl, action => 'create')
          ->name($name . '_create');

        # PUT /article - update article
        $nr->route('/')->via('put')
          ->to(controller => $ctrl, action => 'update')
          ->name($name . '_update');

        # DELETE /article - delete article
        $nr->route('/')->via('delete')
          ->to(controller => $ctrl, action => 'delete')
          ->name($name . '_delete');
        
    }

    # Id passed via URL
    else {
        
        my $nr = $r->route("$path");

        # GET /articles/new - form for create an article
        $nr->route('/new')->via('get')
          ->to(controller => $ctrl, action => 'new_form')
          ->name($name . '_new_form');

        # GET /articles/123 - show article with id 123
        $nr->route('/:id')->via('get')
          ->to(controller => $ctrl, action => 'show')->name($name . '_show');

        # GET /articles/123/delete - form to confirm delete
        $nr->route('/:id/delete')->via('get')
          ->to(controller => $ctrl, action => 'delete_form')
          ->name($name . '_delete_form');

        # GET /articles/123/edit - form for update an article
        $nr->route('/:id/edit')->via('get')
          ->to(controller => $ctrl, action => 'edit_form')
          ->name($name . '_edit_form');

        # GET /articles - list of all articles
        $nr->route('/')->via('get')
          ->to(controller => $ctrl, action => 'index')
          ->name($name . '_index');

        # POST /articles - create new article
        $nr->route('/')->via('post')
          ->to(controller => $ctrl, action => 'create')
          ->name($name . '_create');

        # PUT /articles/123 - update article
        $nr->route('/:id')->via('put')
          ->to(controller => $ctrl, action => 'update')
          ->name($name . '_update');

        # DELETE /articles/123 - delete article
        $nr->route('/:id')->via('delete')
          ->to(controller => $ctrl, action => 'delete')
          ->name($name . '_delete');
    }
}

1;
